src/osd_field.c

```diff
@@ -0,0 +1,359 @@
+{
+	/* custom type identifier: w */
+	if (wl_list_length(&rc.workspace_config.workspaces) > 1) {
+		buf_add(buf, view->workspace->name);
+	}
+}
+
+static void
+field_set_win_state(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: s */
+	if (view->maximized) {
+		buf_add(buf, "M");
+	} else if (view->minimized) {
+		buf_add(buf, "m");
+	} else if (view->fullscreen) {
+		buf_add(buf, "F");
+	} else {
+		buf_add(buf, " ");
+	}
+}
+
+static void
+field_set_win_state_all(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: S */
+	buf_add(buf, view->minimized ? "m" : " ");
+	buf_add(buf, view->maximized ? "M" : " ");
+	buf_add(buf, view->fullscreen ? "F" : " ");
+	/* TODO: add always-on-top and omnipresent ? */
+}
+
+static void
+field_set_output(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: O */
+	if (output_is_usable(view->output)) {
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_output_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: o */
+	if (wl_list_length(&view->server->outputs) > 1 &&
+			output_is_usable(view->output)) {
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_identifier(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: I */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ false));
+}
+
+static void
+field_set_identifier_trimmed(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: i */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ true));
+}
+
+static void
+field_set_title(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: T */
+	buf_add(buf, get_title(view));
+}
+
+static void
+field_set_title_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: t */
+	buf_add(buf, get_title_if_different(view));
+}
+
+static void
+field_set_custom(struct buf *buf, struct view *view, const char *format)
+{
+	if (!format) {
+		wlr_log(WLR_ERROR, "Missing format for custom window switcher field");
+		return;
+	}
+
+	char fmt[LAB_FIELD_SINGLE_FMT_MAX_LEN];
+	unsigned char fmt_len = 0;
+
+	struct buf field_result;
+	buf_init(&field_result);
+
+	char field_formatted[4096];
+
+	for (const char *p = format; *p; p++) {
+		if (!fmt_len) {
+			if (*p == '%') {
+				fmt[fmt_len++] = *p;
+			} else {
+				/*
+				 * Just relay anything not part of a
+				 * format string to the output buffer.
+				 */
+				buf_add_char(buf, *p);
+			}
+			continue;
+		}
+
+		/* Allow string formatting */
+		/* TODO: add . for manual truncating? */
+		if (*p == '-' || *p == '#' || (*p >= '0' && *p <= '9')) {
+			if (fmt_len >= LAB_FIELD_SINGLE_FMT_MAX_LEN - 2) {
+				/* Leave space for terminating 's' and NULL byte */
+				wlr_log(WLR_ERROR,
```

Comment:
Should we continue if we exceed max-len? Might be cleaner to take a more defensive approach and just goto error rather than trying to interpret something that's "undefined".

Reply:
Putting out an error and skipping unknowns might be a better option.

Reply:
I didn't remove the *p == '#', doesn't really do anything as that should never show as a format option.

Reply:
I actually did wind up removing the check for '#' as that denotes an alternate format,
and we don't use that, I also changed the >=0 and <=9 to isdigit()

Todo: check that 1st digit after '%' is not zero, as that says zero fill the output
if '-' is before the '0' then the zero is ignored. But I wasn't sure how much effort 
to put into such a thing.